source/common/upstream/health_discovery_service.cc

```diff
@@ -73,5 +117,86 @@ void HdsDelegate::onRemoteClose(Grpc::Status::GrpcStatus status, const std::stri
   handleFailure();
 }
 
+HdsCluster::HdsCluster(Runtime::Loader& runtime, const envoy::api::v2::Cluster& cluster,
+                       const envoy::api::v2::core::BindConfig& bind_config, Stats::Store& stats,
+                       Ssl::ContextManager& ssl_context_manager,
+                       Secret::SecretManager& secret_manager, bool added_via_api)
+    : info_(new ClusterInfoImpl(cluster, bind_config, runtime, stats, ssl_context_manager,
+                                secret_manager, added_via_api)),
+      runtime_(runtime), initial_hosts_(new HostVector())
+
+{
+  priority_set_.getOrCreateHostSet(0);
+
+  for (const auto& host : cluster.hosts()) {
+    initial_hosts_->emplace_back(HostSharedPtr{new HostImpl(
```

Comment:
s/initial_hosts_/hosts_/ i.e. this is the main data structure you need to maintain in `HdsCluster` as per above comment.